pkg/simulate/pushsync/pushsync.go

```diff
@@ -0,0 +1,304 @@
+
+	// set up clients
+	clients, err := cluster.NodesClients(ctx)
+	if err != nil {
+		return fmt.Errorf("node clients: %w", err)
+	}
+
+	if len(clients) <= 2 {
+		return errors.New("not enough nodes")
+	}
+
+	names := shuffle(rnd, cluster.NodeNames(), o.Seed)
+
+	buckets := ToBuckets(names, o.StartPercentage, o.EndPercentage, o.StepPercentage)
+	uploadNames := buckets[len(buckets)-1]
+
+	malfunctionEth := 0
+
+	for _, bucket := range buckets {
+
+		// STEP: freeze nodes in bucket
+		for _, n := range bucket {
+			node := clients[n]
+
+			nodeAddr, err := node.Addresses(ctx)
+			if err != nil {
+				return err
+			}
+
+			if len(nodeAddr.Underlay) == 0 {
+				return errors.New("not enough underlay addresses")
+			}
+
+			nodeIP := GetIPFromUnderlays(nodeAddr.Underlay)
+
+			fmt.Printf("freezing block number for node %s ip %s\n", n, nodeIP)
+
+			cancelID, err := proxy.Execute(proxyClient.BlockNumberFreeze, nodeIP)
+			if err != nil {
+				return err
+			}
+
+			defer func(cancelID int) {
+				_ = proxy.Cancel(cancelID)
+			}(cancelID)
+		}
+
+		malfunctionEth += len(bucket)
+
+		// Upload nodes are not in the list of malfunctioning nodes
+		// so upload errors should not be expected
+
+		chunks := chunkBatch(rnd, int(o.ChunkCount))
+
+		index := rnd.Intn(len(uploadNames))
+		uploadName := uploadNames[index]
+		uploadNode := clients[uploadName]
+		fmt.Printf("using node %s as uploader\n", uploadName)
+
+		uploaded, err := uploadChunks(ctx, rnd, o, uploadNode, chunks)
+		if err != nil {
+			return fmt.Errorf("upload: %w", err)
+		}
+
+		metricStr := fmt.Sprintf("%d_%d_malfunctioning_backends", malfunctionEth, len(names))
+		metrics.uploadedChunks.WithLabelValues(metricStr).Add(float64(uploaded))
+
+		var downloaded int
+		for i := 0; i < int(o.DownloadCount); i++ {
+			downloadName := randomCmp(rnd, uploadName, names)
+			downloadNode := clients[downloadName]
+
+			fmt.Printf("using node %s as downloader\n", downloadName)
+
+			downloaded = downloadChunks(ctx, o, uploaded, downloadNode, chunks)
+
+			fmt.Printf("%d out of %d_malfunctioning backends\n", malfunctionEth, len(names))
+			fmt.Printf("uploaded to %s %d chunks\n", uploadName, uploaded)
+			fmt.Printf("downloaded from %s %d chunks\n", downloadName, downloaded)
+
+			metrics.downloadCount.WithLabelValues(metricStr).Inc()
+
+			if downloaded == uploaded {
+				break
+			}
+
+			time.Sleep(o.DownloadWait)
+		}
+
+		metrics.downloadedChunks.WithLabelValues(metricStr).Add(float64(downloaded))
+	}
+
+	return nil
+}
+
+func uploadChunks(ctx context.Context, rnd *rand.Rand, o Options, client *bee.Client, chunks []swarm.Chunk) (int, error) {
+
+	batchID, err := client.CreatePostageBatch(ctx, o.PostageAmount, o.PostageDepth, o.GasPrice, "sim-pushsync", false)
+	if err != nil {
+		return 0, fmt.Errorf("batch create %w", err)
+	}
+
+	count := 0
+	for _, chunk := range chunks {
+		_, err := client.UploadChunk(ctx, chunk.Data(), api.UploadOptions{BatchID: batchID})
+		if err == nil {
+			count++
```

Comment:
if the `count` is not equal to `len(chunks)` there's no error... not sure why not to propagate the error here. the upload should pretty much always succeed (at least with our current architecture)

Reply:
Done.